### PR TITLE
Add semantic domains to schema

### PR DESCRIPTION
**Pull request by janetzki**
Semantic domains connect verses with their meaning, which helps to assess a bible revision's quality.

### CODE REVIEW

**Comment by dwhitena** [comment]
@janetzki Looks like there are a couple of conflicts now that we have merged in another change. Could your resolve those? 

@DCNemesis What do you think about about this. Should we somehow merge the idea of Semantic Domains as a Type of annotation with your recent change? Would that make the most sense from your perspective? @janetzki see this PR regarding the "annotations" I'm referring to: https://github.com/BibleNLP/bible-graph/pull/18

**Comment by DCNemesis** [comment]
Hmm ... this is challenging. On the one hand, semantic domains appear to be tied to words (or phrases). But the same word or phrase used in a different sense might not have the same semantic domain. I think I'd prefer that semantic domains be attached as annotations to verses or the actual verse texts.

File: schema.graphql [hunk 1]
```diff
@@ -76,7 +76,7 @@ type VerseText {
   text: String! @search(by: [term])
   verseReference: [VerseReference!]
   bibleRevision: BibleRevision!
-  assessmentDetails: [AssessmentDetail]
```
Comment (janetzki) [inline schema.graphql:79]: #18 renamed `AssessmentDetail` to `TextAnnotation`.

File: schema.graphql [hunk 2]
```diff
@@ -99,7 +99,7 @@ type Assessment {
   bibleRevision: BibleRevision!
   reference: BibleRevision
   date: DateTime! @search
-  details: [Annotation!] @hasInverse(field: assessment)
```
Comment (janetzki) [inline schema.graphql:102]: same here

File: schema.graphql [hunk 3]
```diff
@@ -131,5 +131,24 @@ type ReferenceAnnotation implements Annotation {
...
+  childSemanticDomains: [SemanticDomain] @hasInverse(field: parentSemanticDomain)
+  referenceAnnotations: [ReferenceAnnotation] @hasInverse(field: semanticDomains)
+}
+
+type Word {
+  id: ID!
+  text: String! @search
+  semanticDomains: [SemanticDomain!]
+  isoLanguage: ISOLanguage!
 }
```
Comment (janetzki) [inline schema.graphql:153]: I have added the words to the schema to simplify the reproducibility of the semantic domain tagging. We could also remove them from the schema. What do you prefer, @dwhitena?

**Comment by janetzki** [comment]
> I think I'd prefer that semantic domains be attached as annotations to verses or the actual verse texts.

That sounds reasonable to me, too. As you say, the meaning of a word depends on its context. I have added the semantic domains as an optional part of a `ReferenceAnnotation` and resolved the conflicts. @dwhitena

**Comment by janetzki** [comment]
I marked the PR as a draft because I am experimenting with the schema on DGraph Cloud.

**Comment by DCNemesis** [comment]
I almost wonder if the Semantic Domain stuff can be built up and linked in as a Federated Graph using Apollo? (https://www.apollographql.com/blog/announcement/apollo-federation-f260cf525d21/) The annotations concept is built on the idea that some of the details of the specific methods may be documented elsewhere. But if that "documentation" is also best represented as a graph ... maybe a federated graph makes sense? I do not know whether a graph can be federated across services ... like the core graph on dgraph and the rest somewhere else.

**Comment by janetzki** [comment]
The PR is ready for review. I added a node type for `SemanticDomainQuestion` as an intermediary layer between `SemanticDomain` and `Word`, which allows a more precise semantic hierarchy of `Words`.

**Examples**
`SemanticDomain`: "1.6.3.1 Egg"
`SemanticDomainQuestion`: "What types of eggs are there?"
`Words`: ["bird egg", "reptile egg", "insect egg"]

**Comment by janetzki** [comment]
> I almost wonder if the Semantic Domain stuff can be built up and linked in as a Federated Graph using Apollo?

Interesting idea. Our schema is getting large, and I did not find a way to modularize it in Dgraph Cloud. For now, federating the graph (in Apollo or across services) seems like an overhead to me, but it might become beneficial if the schema grows further. What do you think, @dwhitena?

**Comment by dwhitena** [comment]
Great discussion here @janetzki  and @DCNemesis. My general rule here is to optimize when we need to optimize for a specific reason. Rather than adding complication based on an "anticipated" scaling issue later on. So I would keep things together for now until we are hitting a wall.

**Comment by DCNemesis** [comment]
@janetzki have you migrated the Dgraph instance you have to this proposed version without issue? If so, I think we can approve this PR. I've been waiting to apply some of the schema updates to our own instance until I get through some work that needs to use the graph a lot.

**Comment by janetzki** [comment]
> have you migrated the Dgraph instance you have to this proposed version without issue?

@DCNemesis Yes, our SIL-AI/hpi-dev Dgraph instance uses this schema and works.

**Review by DCNemesis — APPROVED** [review]

**Pull request merged**